### ISSUE TITLE
Incompatiblity with pnpm >11's virtualStoreType: global

### ISSUE THREAD

**Issue by bessey** [oxc-project/tsgolint#1168]
## Summary

With pnpm's global virtual store (`virtualStoreType: global`), `oxlint --type-aware`
resolves some packages' exported types as `any`, while `tsc` on the same
`tsconfig.json` types them correctly. Turning the setting off makes oxlint agree with
`tsc` again, with no other change.

Because the degraded type is `any`, type-aware rules then fire incorrectly:

- `typescript/no-redundant-type-constituents` reports `'any' overrides all other types`
  for any intersection or union built from the affected types.
- `typescript/no-unnecessary-type-assertion` reports assertions as unnecessary. This one
  is actively harmful: following its advice and removing the assertion makes `tsc` fail.

No diagnostic is emitted about the failed resolution itself, so the only visible symptom
is a confusing lint error in first-party code that is actually correct.

## Environment

|                 |                                    |
| --------------- | ---------------------------------- |
| oxlint          | 1.72.0                             |
| oxlint-tsgolint | 0.24.0 (also reproduced on 0.25.0) |
| typescript      | 7.0.2 (native)                     |
| pnpm            | 12.2.1                             |
| node            | v24.18.1                           |
| OS              | macOS 26.6.2, arm64                |

pnpm workspace settings relevant to the repro:

```yaml
virtualStoreType: global
```

The consuming package's `tsconfig.json`:

```json
{
  "extends": "expo/tsconfig.base",
  "compilerOptions": {
    "strict": true,
    "moduleResolution": "bundler",
    "isolatedModules": true,
    "composite": true,
    "types": ["node"]
  },
  "include": ["src/**/*", "config/**/*", "App.tsx"]
}
```

With `virtualStoreType: global`, packages resolve outside the project directory:

```
~/Library/pnpm/store/v11/links/@gluestack-ui/themed/1.1.73/<hash>/node_modules/@gluestack-ui/themed
```

## Reproduction

In a React Native app depending on `@gluestack-ui/themed@1.1.73` and
`react-native@0.85.3`, with `virtualStoreType: global` set:

```ts
import type { ComponentPropsWithoutRef } from "react";
import { Box } from "@gluestack-ui/themed";

type IsAny<T> = 0 extends 1 & T ? true : false;
type AssertFalse<T extends false> = T;

// tsc errors here only if the props type IS any
export type NotAny = AssertFalse<IsAny<ComponentPropsWithoutRef<typeof Box>>>;

// oxlint errors here only if the right-hand side IS any
export type Intersected = { a: 1 } & ComponentPropsWithoutRef<typeof Box>;
```

Then:

```
$ pnpm exec oxlint --type-aware --quiet src/probe.ts
src/probe.ts:12:38: error typescript(no-redundant-type-constituents): 'any' overrides all other types in this intersection type.

$ pnpm exec tsc --noEmit
(clean)
```

**Expected:** both agree. Either the props type is `any`, in which case `tsc` should fail
on `NotAny`, or it is not, in which case oxlint should not fail on `Intersected`.

**Actual:** oxlint treats it as `any`; `tsc` does not. Reproduces identically in `.ts`
and `.tsx` files.

Removing `virtualStoreType: global` and reinstalling makes oxlint clean, with no source
change.

## Scope — it is not every package

Same file, same program, same store. Only some packages degrade:

| Import                             | Types resolve under global store |
| ---------------------------------- | -------------------------------- |
| `react-native`                     | yes                              |
| `@tanstack/react-query`            | yes                              |
| `date-fns`                         | yes                              |
| `@gluestack-ui/themed`             | **no — `any`**                   |
| `react-native-keyboard-controller` | **no — `any`**                   |

The two failing packages have `react-native` as a `peerDependency` and reference its
types from their own shipped `.d.ts` files. The passing ones either have no external type
imports or import only packages present in the project's own `node_modules`. That
suggests the failure is in resolving a **dependency of a `.d.ts` file** when the importing
package's realpath lies outside the project, rather than in resolving the entry package
itself — but we have not confirmed the exact failing resolution step, and `tsc` resolves
the same graph without difficulty.

## Ruled out

To save time — each of these was tested and is not the cause:

- **Missing implicit dependency.** For `@gluestack-ui/themed`, `react-native` _is_ present
  as a sibling in its store directory, complete with its `types/` directory. Declaring it
  explicitly via pnpm `packageExtensions` (`peerDependencies: react-native: '*'`), across
  the whole 507-package closure, changes nothing.
- **Duplicate copies.** There is exactly one `react-native` in the closure. The copy the
  app resolves and the copy sibling to `@gluestack-ui/themed` are the same realpath.
- **Incomplete store.** Each affected package's declared `types` entry file exists and is
  readable at its store path.
- **A TypeScript regression.** Reproduced with the repo's TypeScript 7.0.2 and separately
  with TypeScript 5.9.3 — `tsc` is correct in both, so the disagreement is oxlint-side.
- **oxlint-tsgolint version.** 0.25.0 produces identical results to 0.24.0.
- **pnpm `publicHoistPattern`.** Adding the affected scope to it has no effect (these are
  direct dependencies, which public hoisting does not affect).

## Impact

`virtualStoreType: global` is attractive in a monorepo with many git worktrees — it
replaces a full per-worktree `node_modules` with links into one shared store. 

## Workaround

Do not set `virtualStoreType: global`, or disable the affected rules per file.

**Comment by camc314** [comment]
Hi! Thanks for reporting this issue.

I reproduced the `Box` example with the global virtual store. In  my reproduction, `react-native` resoled, but `@gluestack-ui/themed` cannot resolve React's declarations (`@types/react`) from its store location.

Expo's base config includes `skipLibCheck` which hides the unerlying TS7016 error in `Box/index.d.ts`.

Running this exposes the error:

```sh
pnpm exec tsc --noEmit --skipLibCheck false
```

The `IsAny` assertion unfortunately doesn’t rule this out: TypeScript propagates an internal error type through the conditional, so `NotAny` doesn’t resolve to the expected boolean.

This [[package extension](https://github.com/pnpm/pnpm.io/blob/main/versioned_docs/version-10.x/settings.md)](https://github.com/pnpm/pnpm.io/blob/main/versioned_docs/version-10.x/settings.md) fixes the reported lint error in my reproduction while retaining the global store:

```yaml
# pnpm-workspace.yaml
virtualStoreType: global

packageExtensions:
  '@gluestack-ui/themed@1.1.73':
    peerDependencies:
      '@types/react': '*'
```

Keep your project’s compatible `@types/react` installed, then run `pnpm install`.

Could you try this against your project? I haven’t verified `react-native-keyboard-controller`; it may need a separate dependency correction. If this resolves the Box case, the upstream fix would be for the package to declare its React type dependency.

**Comment by bessey** [comment]
It turns out this is a now known issue with pnpm's `virtualStoreType: global`. oxc is innocent: https://github.com/pnpm/pnpm/issues/12437
